### PR TITLE
Add autologging safety utils to pytorch integration

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <17039389+harupy@users.noreply.github.com>

## What changes are proposed in this pull request?

Add autologging safety utils to pytorch integration

## How is this patch tested?

Existing unit tests.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/pytorch/_pytorch_autolog.py [hunk 1]
```diff
@@ -45,12 +53,14 @@ def _autolog(log_every_n_epoch=1, log_models=True):
...
                        If ``False``, trained models are not logged.
+    :param disable: If ``True``, disables all supported autologging integrations. If ``False``,
+                    enables all supported autologging integrations.
     """
     global every_n_epoch
     every_n_epoch = log_every_n_epoch
 
     def getPLCallback(log_models, metrics_logger):
-        class __MLflowPLCallback(pl.Callback):
+        class __MLflowPLCallback(pl.Callback, metaclass=ExceptionSafeClass):
```
Comment (harupy) [inline mlflow/pytorch/_pytorch_autolog.py:63]: This line causes the following error:

```
TypeError: metaclass conflict: the metaclass of a derived class must be a (non-strict) subclass
of the metaclasses of all its bases
```

`pl.Callback` is an abstract class.

Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py:63]: Found a workaround: https://stackoverflow.com/a/28727066/6943581

Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py:65]: A workaround:

```python
        class Callback:
            def __getattr__(self, name):
                return pl.Callback().__getattribute__(name)

        class __MLflowPLCallback(Callback, metaclass=ExceptionSafeClass):
```

Comment (dbczumar) [reply mlflow/pytorch/_pytorch_autolog.py:65]: @harupy Is it possible to generate `ExceptionSafeClass` based on a base type? E.g., write a function that provides the metaclass given base type `type` or base type `abc.ABC`?

Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py:65]: Like this?

```python
def create_ExceptionSafeClass(base):
    class ExceptionSafeClass(base):
        ...

    return ExceptionSafeClass
```

Comment (dbczumar) [reply mlflow/pytorch/_pytorch_autolog.py:65]: Yeah, does that work?

Comment (dbczumar) [reply mlflow/pytorch/_pytorch_autolog.py:65]: For simplicity, assuming there aren’t that many bases used, we can define a couple of these statically. E.g ExceptionSafeClass = function (type) and ExceptionSafeAbstractClass = function(abc.ABC)

Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py:65]: Tried this

```python
class ExceptionSafeAbstractClass(abc.ABC):
    def __new__(cls, name, bases, dct):
        for m in dct:
            if callable(dct[m]):
                dct[m] = exception_safe_function(dct[m])
        return type.__new__(cls, name, bases + (abc.ABC,), dct)

__MLflowPLCallback(Callback, metaclass=ExceptionSafeAbstractClass)
```

but didn't work.

Comment (dbczumar) [reply mlflow/pytorch/_pytorch_autolog.py:65]: @harupy This seems to work:

```
import abc
class ExceptionSafeAbstractClass(abc.ABCMeta):

    def __new__(mcls, name, bases, namespace, **kwargs):
        for m in namespace:
            if callable(namespace[m]):
                print(namespace[m])
                namespace[m] = exception_safe_function(namespace[m])
        return abc.ABCMeta.__new__(mcls, name, bases, namespace, **kwargs)
```

Can we use this and update the validation logic for `safe_patch` to assert that the metaclass is either `ExceptionSafeClass` or `ExceptionSafeAbstractClass`?

Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py]: @dbczumar 

Should we update the relevent tests?

```diff
 @pytest.mark.usefixtures(test_mode_on.__name__)
-def test_exception_safe_class_exhibits_expected_behavior_in_test_mode():
+@pytest.mark.parametrize(
+    "baseclass,metaclass", [(object, ExceptionSafeClass), (abc.ABC, ExceptionSafeAbstractClass)]
+)
+def test_exception_safe_class_exhibits_expected_behavior_in_test_mode(baseclass, metaclass):
     assert autologging_utils._is_testing()
 
-    class NonThrowingClass(metaclass=ExceptionSafeClass):
+    class NonThrowingClass(baseclass, metaclass=metaclass):
         def function(self):
             return 10
 
@@ -497,7 +504,7 @@ def test_exception_safe_class_exhibits_expected_behavior_in_test_mode():
 
     exc_to_throw = Exception("function error")
 
-    class ThrowingClass(metaclass=ExceptionSafeClass):
+    class ThrowingClass(baseclass, metaclass=metaclass):
         def function(self):
             raise exc_to_throw
```

**Comment by harupy** [comment]
pylint complains:

```
========== pylint ==========

Using config file /home/runner/work/mlflow/mlflow/pylintrc
************* Module mlflow.gluon
mlflow/gluon.py (332,8): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
************* Module mlflow.keras
mlflow/keras.py (623,8): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
************* Module mlflow.pytorch._pytorch_autolog
mlflow/pytorch/_pytorch_autolog.py (65,8): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
************* Module tests.autologging.test_autologging_safety_unit
tests/autologging/test_autologging_safety_unit.py (466,4): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
tests/autologging/test_autologging_safety_unit.py (474,4): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
tests/autologging/test_autologging_safety_unit.py (492,4): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
tests/autologging/test_autologging_safety_unit.py (500,4): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used
tests/autologging/test_autologging_safety_unit.py (741,4): [E1139 invalid-metaclass] Invalid metaclass '_ExceptionSafeClass' used

```

**Comment by harupy** [comment]
Got this error:

https://github.com/mlflow/mlflow/pull/3855/checks?check_run_id=1563229497#step:4:162

```
E           AttributeError: <module 'mlflow.pytorch' from '/home/runner/work/mlflow/mlflow/mlflow/pytorch/__init__.py'> does not have the attribute 'safe_patch'
```


In pytorch autlogging, `safe_patch` is imported & called in `mlflow/pytorch/_pytorch_autolog.py`.

File: mlflow/pytorch/_pytorch_autolog.py (hunk 1 above)
Comment (dbczumar) [reply mlflow/pytorch/_pytorch_autolog.py]: Yes :)

**Comment by dbczumar** [comment]
> Got this error:
> 
> https://github.com/mlflow/mlflow/pull/3855/checks?check_run_id=1563229497#step:4:162
> 
> ```
> E           AttributeError: <module 'mlflow.pytorch' from '/home/runner/work/mlflow/mlflow/mlflow/pytorch/__init__.py'> does not have the attribute 'safe_patch'
> ```
> 
> In pytorch autlogging, `safe_patch` is imported & called in `mlflow/pytorch/_pytorch_autolog.py`.
> 
> I think we can fix this by peforming patching in `mlflow/pytorch/__init__.py` instead of `mlflow/pytorch/_pytorch_autolog.py`.
> 
> ```python
> # mlflow/pytorch/_pytorch_autolog.py
> 
> def _create_patch_fit(...):
>     def fit(...):
>         ...
>     return fit
> 
> # ------------------------
> 
> # mlflow/pytorch/__init__.py
> 
> def autolog(...):
>     from mlflow.pytorch._pytorch_autolog import _create_patch_fit
> 
>     patch_fit = _create_patch_fit()
>     safe_patch(FLAVOR_NAME, pl.Trainer, "fit", patch_fit, manage_run=True)
> ```

Sounds good to me!

File: mlflow/pytorch/_pytorch_autolog.py [hunk 2]
```diff
@@ -30,8 +33,7 @@
 
 
 @rank_zero_only
-@experimental
-def _autolog(log_every_n_epoch=1, log_models=True):
+def _create_patch_fit(log_every_n_epoch=1, log_models=True):
     """
     Enable automatic logging from pytorch to MLflow.
     Logs loss and any other metrics specified in the fit
```
Comment (harupy) [inline mlflow/pytorch/_pytorch_autolog.py:39]: I'll update docstring.

File: mlflow/pytorch/_pytorch_autolog.py (hunk 1 above)
Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py]: done!

File: mlflow/pytorch/_pytorch_autolog.py (hunk 2 above)
Comment (harupy) [reply mlflow/pytorch/_pytorch_autolog.py]: done.

File: mlflow/utils/autologging_utils.py [hunk 3]
```diff
@@ -406,26 +407,56 @@ def safe_function(*args, **kwargs):
     return safe_function
 
 
-class ExceptionSafeClass(type):
+def _exception_safe_class_factory(base_class):
```
Comment (harupy) [inline mlflow/utils/autologging_utils.py:410]: Found another approach:


```python
def _exception_safe_class_factory(base_class, name):
    docstring = """
...
"""

    def __new__(cls, name, bases, dct):
        for m in dct:
            if callable(dct[m]):
                dct[m] = exception_safe_function(dct[m])
        return base_class.__new__(cls, name, bases, dct)

    clazz = type(name, (base_class,), {"__new__": __new__})
    clazz.__doc__ = docstring

    return clazz
```

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @harupy !

**Pull request merged**